core/src/Pos/Core/Slotting/EpochOrSlot.hs

```diff
@@ -44,6 +44,10 @@ import           Pos.Core.Slotting.SlotId
 
 -- | Represents SlotId or EpochIndex. Useful because genesis blocks
 -- have only EpochIndex, while main blocks have SlotId.
+
+-- This type is a pain in the neck, but we have to live with it forever
+-- because this encoding is baked into the system. The best we can do is
+-- provide helper functions to minimise the pain.
```

Comment:
Can you explain the problems with using the type a bit further (here rather than in the code). How is it baked into the system? Couldn't we have an alternative in Haskell-land that is encoded the same way in the blocks?

Reply:
Further comments added.